[10기 Tami] TodoList with CRUD

ink-0 opened the pull request:
### 📄 회고
기타 사정으로 인해 요구사항을 거의 완료하지 못했습니다. 😢
이번주 내내 스피드를 내서 이번주 미션과 저번주 미션을 이행할 예정입니다

### 📌 진행 상황
- html 모듈화
- 체크 박스 구현 중에 상태관리에 있어 헷갈리는 부분이 있습니다.


---
### 🎯 요구사항
- [x] todo list에 todoItem을 키보드로 입력하여 추가하기
- [x] todo list의 체크박스를 클릭하여 complete 상태로 변경. (li tag 에 completed class 추가, input 태그에 checked 속성 추가)
- [x] todo list의 x버튼을 이용해서 해당 엘리먼트를 삭제
- [x] todo list를 더블클릭했을 때 input 모드로 변경. (li tag 에 editing class 추가) 단 이때 수정을 완료하지 않은 상태에서 esc키를 누르면 수정되지 않은 채로 다시 view 모드로 복귀
- [x] todo list의 item갯수를 count한 갯수를 리스트의 하단에 보여주기
- [x] todo list의 상태값을 확인하여, 해야할 일과, 완료한 일을 클릭하면 해당 상태의 아이템만 보여주기

ink-0 committed:
[week1] TodoUI UI component 분리

-ink-0/js-todo-list-step1/#1

ink-0 committed:
[week1] favicon 변경

-ink-0/js-todo-list-step1/#1

ink-0 committed:
[Week1] TodoInput 기능 구현현

- ink-0/js-todo-list-step1/#1

ink-0 committed:
TodoCount UI 분리

ink-0/js-todo-list-step1/#1

ink-0 committed:
TodoList UI 수정

ink-0/js-todo-list-step1/#1

ink-0 committed:
Merge pull request #2 from ink-0/week1/1/TodoUI

[Week1] TodoList UI 모듈화

ink-0 committed:
[week1] todoes data 변경 (idx,content 추가)

-ink-0/js-todo-list-step1/#3

ink-0 committed:
[week1] checkBox 기능생성중

-ink-0/js-todo-list-step1/#3

ink-0 committed:
Merge pull request #4 from ink-0/week1/3/checkBox

Week1/3/check box

auszlee commented on index.js:
```diff
@@ -0,0 +1,2 @@
+import TodoApp from './src/js/TodoApp.js';
+new TodoApp(document.querySelector('.App'));
```
main 전체를 넘겨주는 건 생각 못해봤는데 HTML이 한결 더 깔끔해지네요! 하나 배워갑니다 👍

renee13wi commented on src/js/TodoApp.js:
```diff
@@ -0,0 +1,63 @@
+  new TodoInput({
+    $app,
+    onAdd: (contents) => {
+      const prevIdx = this.state.todoes[this.state.todoes.length - 1].idx;
+
+      const newTodo = {
+        idx: prevIdx + 1,
+        content: contents,
+        state: '',
+        edit: '',
+      };
+      this.state.todoes.push(newTodo);
+
+      this.setState({
+        ...this.state,
+      });
+    },
```
이 부분은 따로 함수를 만들어 주는게 좋을 것 같아요! :)
```ex) handleAdd```

ink-0 replied:
네이밍까지 감사합니다 ㅎㅎ

renee13wi commented on src/js/TodoApp.js:
```diff
@@ -0,0 +1,63 @@
+    todoCount.setState(this.state.todoes);
+  };
+
+  new TodoInput({
```
```
const todoList = new TodoList({
    $app,
    initialState: this.state.todoes,
  });
  const todoCount = new TodoCount({
    $app,
    initialState: this.state.count,
  });
```
TodoList, TodoCount와 다르게 인스턴스를 생성 할 때 const에 할당을 안해준 이유가 있을까요?!

ink-0 replied:
일단은 할당한 후에 따로 `setState`할 필요가 없다고 생각해 그렇게 했는데요, 앞으로 추가 기능을 구현하면서 어떻게 될지는 모르겠네요 ..^^

renee13wi commented on src/js/TodoCount.js:
```diff
@@ -0,0 +1,28 @@
+
+  this.render = () => {
+    const todoCountTemplate = `
+      <span class="todo-count">총 <strong>${this.state}</strong> 개</span><ul class="filters">
```
```<ul class="filters">```
가독성을 위해 개행해주는게 좋을 것 같아요! 😄

ink-0 replied:
꼼꼼한 리뷰 감사합니다

renee13wi reviewed (commented):
Tami님, 1주차 과제 너무 수고많으셨습니다 😄 
1주차에 PR 올리셨단 것 만으로도 대단하십니다! 😄 

남은 과제도 함께 열심히 해봅시다! 화이팅 !!

auszlee commented on src/js/TodoItem.js:
```diff
@@ -0,0 +1,21 @@
+    this.state = nextState;
+    this.render();
+  };
+  this.render = () => {
```
컴포넌트마다 render를 따로 둔 이유가 있으신가요? 그렇게는 아직 못해봐서 ㅎㅎㅎ 궁금하네요

ink-0 replied:
사실 제가 다른 상태관리 블로그를 많이 참고해서 만드느라 비슷해 졌는데 eazisilver 님의 코드도 한번 보고 다른점을 찾아보면 좋을 것 같습니다 ㅎㅎ

auszlee reviewed (commented):
안녕하세요 이지은입니다.
코드리뷰가 처음이라 다른사람 코드보는게 익숙하지가 않네요 😂
기능구현은 아직 다 안됐어도 컴포넌트가 잘 나눠져있어서 리뷰하면서 많이 배웠습니다

nerdyinu commented on src/js/TodoCount.js:
```diff
@@ -0,0 +1,28 @@
+  this.state = initialState;
+
+  this.$target = document.createElement('div');
+  this.$target.className = 'count-container';
```
이렇게 새로 createElement를 하는 방식보단 TodoApp의 template에서 하위 component들에 해당하는 target을 배분해주는 방식은 어떨까요?

nerdyinu reviewed (commented):
잘 읽었습니다! 전체적으로 이해하기 쉽게 잘 짜여져 있는 것 같은데 templated을 각 컴포넌트에서 따로따로 생성하는 방식보다는 TodoApp에서 각 컴포넌트의 target에 해당하는 엘리먼트를 template으로 render하고 각 컴포넌트 객체 생성과 함께 target값을 넘겨주는 방식은 어떨까요? 코드 추상화에 신경을 좀 더 쓰신 다면 더 좋은 코드가 나올 것 같아요!

nerdyinu commented on src/js/TodoApp.js:
```diff
@@ -0,0 +1,63 @@
+  const todoCount = new TodoCount({
+    $app,
+    initialState: this.state.count,
+  });
```
TodoList나 TodoCount에 사용하실 함수들도  OnAdd처럼 객체에 화살표 함수 형식으로 넘겨주거나, 아니면 bind(this)를 해서 넘겨주면 객체에서 state 충돌없이 사용가능해요!

ink-0 committed:
[week1] checkBox 구현

-ink-0/js-todo-list-step1/#3

ink-0 committed:
Merge pull request #5 from ink-0/week1/3/checkBox

[week1] checkBox 구현

ink-0 committed:
[week1] TodoList Delete 구현

- ink-0/js-todo-list-step1/#6

ink-0 committed:
Merge pull request #7 from ink-0/week1/6/TodoDelete

[week1] TodoList Delete 구현

ink-0 committed:
[week1] TodoList Edit 구현

- ink-0/js-todo-list-step1/#8

ink-0 committed:
[week1] TodoList Edit 구현 일부 수정

- ink-0/js-todo-list-step1/#8

ink-0 committed:
[week1] TodoList 리팩토링

- ink-0/js-todo-list-step1/#8

ink-0 committed:
Merge pull request #9 from ink-0/week1/8/TodoEdit

Week1/8/todo edit

ink-0 committed:
[week1] TodoFilter 구현

- 상태 추가 filterTodoes, isFilter, todoesCount
- component 네이밍 변경  TodoCount ->  TodoFilter
- ink-0/js-todo-list-step1/#10

ink-0 committed:
[week1] 코드 리팩토링
- const 생성: 메세지, class data 분리
- 객체 리터럴 적용
- ink-0/js-todo-list-step1/#10

ink-0 committed:
Merge pull request #11 from ink-0/week1/10/TodoFilter

Week1/10/todo filter

ink-0 committed:
[week1] TodoFilter CSS 조건 수정

- 그외 const 대치 안한 것 수정
- ink-0/js-todo-list-step1/#12

ink-0 committed:
[week1] localStorage 적용

-ink-0/js-todo-list-step1/#12

ink-0 committed:
Merge pull request #13 from ink-0/week1/12/localStorage

Week1/12/local storage

ink-0 committed:
[week2] GET userList

-ink-0/js-todo-list-step1/#14

ink-0 committed:
Merge pull request #16 from ink-0/week2/14/GETuser

[week2] GET userList

ink-0 committed:
[week2] UserList UI 생성
]
-ink-0/js-todo-list-step1/#17

ink-0 committed:
[week2] UUserList 의 user에 클릭이벤트 거는중
]
-ink-0/js-todo-list-step1/#17

ink-0 committed:
Merge pull request #18 from ink-0/week2/17/UserListUI

Week2/17/user list UI

ink-0 committed:
chore:

ink-0 commented on src/js/TodoApp.js:
```diff
@@ -0,0 +1,179 @@
+import getUserData from './core/getUserData.js';
+
+export default function TodoApp($app) {
+  const initialData = {
+    todoes: [
+      {
+        idx: 0,
+        content: 'hiEvery One',
+        state: '',
+      },
+      {
+        idx: 1,
+        content: 'Im Tami',
+        state: '',
+      },
+    ],
+    todoesFiltered: [],
+    filterState: FILTER_TYPES.ALL,
+    todoesCount: '0',
+    users: [],
+  };
```
초기 데이터를 따로 보관하는 것이 좋았을 것이다.

ink-0 commented on src/js/TodoApp.js:
```diff
@@ -0,0 +1,179 @@
+    const resetTodoes = todoes.filter((todo) => {
+      return todo.idx !== idx;
+    });
+    this.state['todoes'] = resetTodoes;
+    localStorage.clear();
+    localStorage.setItem('state', JSON.stringify({ ...this.state }));
+    this.state = JSON.parse(localStorage.getItem('state'));
+    this.setState({
+      ...this.state,
+    });
+  };
```
생성 삭제 시 `clear` 후 set하는 방식밖에 없었는지 고민할 필요가 있다.

ink-0 commented on src/js/TodoApp.js:
```diff
@@ -0,0 +1,179 @@
+    });
+  };
+
+  const filterTodo = (filterType) => {
+    const { todoes } = this.state;
+    if (filterType === FILTER_TYPES.ALL) {
+      this.setState({
+        ...this.state,
+        filterState: FILTER_TYPES.ALL,
+        todoesCount: todoes.length,
+      });
+    } else if (filterType === FILTER_TYPES.COMPLETE) {
+      const completedTodoes = todoes.filter(
+        (todo) => todo.state === FILTER_TYPES.COMPLETE
+      );
+
+      this.setState({
+        ...this.state,
+        todoesFiltered: completedTodoes,
+        filterState: FILTER_TYPES.COMPLETE,
+        todoesCount: completedTodoes.length,
+      });
+    } else if (filterType === FILTER_TYPES.ACTIVE) {
+      const activeTodoes = todoes.filter(
+        (todo) => todo.state !== FILTER_TYPES.COMPLETE
+      );
+      this.setState({
+        ...this.state,
+        todoesFiltered: activeTodoes,
+        filterState: FILTER_TYPES.ACTIVE,
+        todoesCount: activeTodoes.length,
+      });
```
코드 중복이 많다.
setState를 하는 부분에서 중복을 해결할 방안이 필요하다

ink-0 commented on src/js/components/TodoList.js:
```diff
@@ -0,0 +1,100 @@
+    this.render();
+  };
+
+  $nodeTodoList.addEventListener('click', (e) => {
+    const $node = e.target;
+
+    if ($node.className === TODO_ITEM_CLASS.TOGGLE) {
+      this.toggleTodoItem($node);
+    }
+    if ($node.className === TODO_ITEM_CLASS.DESTROY) {
+      this.deleteTodoItem($node);
+    }
+
+    if ($node.className === TODO_ITEM_CLASS.EDIT) {
+      this.editTodoItem($node);
+    }
```
이벤트 early return 이 필요하다

ink-0 reviewed (commented):
1차 TodoList 본인 리뷰
